llvm_objects.py

```diff
@@ -0,0 +1,238 @@
+
+    def get_instructions(self, l_block, w_module):
+        l_instr = LLVMGetFirstInstruction(l_block)
+        w_instrs = []
+        prev = None
+        while l_instr:
+            current_instr = W_Instruction(l_instr, w_module)
+            if prev:
+                prev.w_next_instr = current_instr
+            prev = current_instr
+            w_instrs.append(current_instr)
+            l_instr = LLVMGetNextInstruction(l_instr)
+        return w_instrs
+
+    def get_first_instruction(self):
+        return self.w_instructions[0]
+
+    def __eq__(self, other):
+        if isinstance(other, W_Block):
+            return self.w_instructions == other.w_instructions and\
+                   self.l_value == other.l_value
+        return NotImplemented
+
+class W_Instruction(object):
+    ''' Represents a wrapper for an LLVMValueRef that is an instruction. '''
+
+    def __init__(self, l_instr, w_module):
+        self.addr = rffi.cast(rffi.INT, l_instr)
+        self.l_operands = self.get_operands(l_instr, w_module)
+        self.opcode = LLVMGetInstructionOpcode(l_instr)
+        self.w_next_instr = None
+        if self.opcode == LLVMBr:
+            self.is_conditional = LLVMIsConditional(l_instr)
+            # if a branch instruction is conditional, store the condition
+            # and the two basic blocks to jump to if it is true or false
+            if self.is_conditional:
+               self.condition = LLVMGetCondition(l_instr)
+               self.l_bb_true = LLVMValueAsBasicBlock(self.l_operands[2])
+               self.l_bb_false = LLVMValueAsBasicBlock(self.l_operands[1])
+            # otherwise, store the basic block to jump to unconditionally
+            else:
+                self.l_bb_uncond = LLVMValueAsBasicBlock(self.l_operands[0])
+        elif self.opcode == LLVMICmp:
+            self.icmp_predicate = LLVMGetICmpPredicate(l_instr)
+        elif self.opcode == LLVMPHI:
+            self.count_incoming = LLVMCountIncoming(l_instr)
+            self.incoming_block = []
+            self.incoming_value = []
+            for i in range(self.count_incoming):
+                self.incoming_block.append(LLVMGetIncomingBlock(l_instr, i))
+                self.incoming_value.append(LLVMGetIncomingValue(l_instr, i))
+        elif self.opcode == LLVMCall:
+            if not LLVMIsDeclaration(self.l_operands[-1]):
+                self.func_param_count = LLVMCountParams(self.l_operands[-1])
+                self.l_func_params = []
+                for index in range(self.func_param_count):
+                    self.l_func_params.append(LLVMGetParam(self.l_operands[-1], index))
+            else:
+                # assume the instruction is either printf or puts (this is checked by the
+                # the interpreter)
+                self.l_string_format_ref = LLVMGetOperand(self.l_operands[0], 0)
+                self.func_name = rffi.charp2str(LLVMGetValueName(self.l_operands[-1]))
```

Comment:
All these attributes are specific to the type of instruction. So it would probably be better to have specific instruction classes for them, because otherwise every (for example) addition will have space for a ``l_bb_true`` field.